src/main.rs

```diff
@@ -1,7 +1,160 @@
+            is_running: true,
+            stdout: BufWriter::new(io::stdout()),
+        })
+    }
+
+    pub fn run(&mut self) -> Result<()> {
+        let _override = RawModeOverride::new()?;
+
+        while self.is_running {
+            self.render()?;
+            self.process_input()?;
+        }
+
+        Ok(())
+    }
+
+    fn terminate(&mut self) -> Result<()> {
+        self.is_running = false;
+        queue!(self.stdout, Clear(All), MoveTo(0, 0))?;
+        Ok(())
+    }
+
+    fn render(&mut self) -> Result<()> {
+        queue!(self.stdout, Hide)?;
+        queue!(self.stdout, MoveTo(0, 0))?;
+
+        self.render_rows()?;
+
+        queue!(self.stdout, Show)?;
+        queue!(self.stdout, MoveTo(self.cursor.x, self.cursor.y))?;
+
+        self.stdout.flush()?;
+        Ok(())
+    }
+
+    fn render_rows(&mut self) -> Result<()> {
+        for row in 0..self.dimentions.height {
+            if row == self.dimentions.height / 3 {
+                let message_row = self.make_message_row();
+                queue!(self.stdout, Print(message_row))?;
+            } else {
+                queue!(self.stdout, Print("~"))?;
+            }
+
+            queue!(self.stdout, Clear(UntilNewLine))?;
+            if row < self.dimentions.height - 1 {
+                queue!(self.stdout, Print("\r\n"))?;
+            }
+        }
+
+        Ok(())
+    }
+
+    fn make_message_row(&self) -> String {
+        let message = "kilo-rs (week 2)";
+        let width_bound = std::cmp::min(message.len(), self.dimentions.width as usize);
+        let message = &message[..width_bound];
+
+        let padding_len = (self.dimentions.width as usize - message.len()) / 2;
+        if padding_len > 0 {
+            format!("~{}{message}", " ".repeat(padding_len - 1))
+        } else {
+            String::from(message)
+        }
+    }
+
+    #[rustfmt::skip]
+    fn process_input(&mut self) -> Result<()> {
+        if let Event::Key(KeyEvent { code, modifiers }) = event::read()? {
+            use event::KeyCode::*;
+            use event::KeyModifiers as KM;
+
+            match (modifiers, code) {
+                (KM::NONE, Up)       => self.move_cursor_up(),
+                (KM::NONE, Down)     => self.move_cursor_down(),
+                (KM::NONE, Left)     => self.move_cursor_left(),
+                (KM::NONE, Right)    => self.move_cursor_right(),
+
+                (KM::NONE, Home)     => self.move_cursor_to_line_start(),
+                (KM::NONE, End)      => self.move_cursor_to_line_end(),
+
+                (KM::NONE, PageUp)   => self.move_cursor_to_screen_top(),
+                (KM::NONE, PageDown) => self.move_cursor_to_screen_bottom(),
+
+                (KM::CONTROL, Char('q')) => self.terminate()?,
+
+                _ => {},
+            }
+        }
+
+        Ok(())
+    }
+
+    fn move_cursor_up(&mut self) {
+        if self.cursor.y > 0 {
+            self.cursor.y -= 1;
```

Comment:
Either use saturating_sub or max. Also consider creating methods in the cursor struct directly, eg. `up(y_delta)` to reduce responsibility on editor class. Further block public access to the internal x/y members.

Reply:
1. To me using `saturating_sub` doesn't sound right. It's there to deal with overflow, whereas I'm describing logic (the cursor can't move into negative coords). I wouldn't mind using `max`, but I don't see a good way of doing that since `y` is unsigned and overflow gets in the way again.
2. I'm treating the cursor struct as POD. Therefore all fields are public. In general I'm against putting movement methods on the cursor. I believe it should be a responsibility of some other entity higher up the abstraction tree.